compose: initial file watch docs (#17010)
docker/docs: compose/file-watch.md

Each rule requires, a `path` pattern and `action` to take when a modification is detected. There are two possible actions for `watch` and depending on
the `action`, additional fields might be accepted or required. 

### `action`

#### Sync

If `action` is set to `sync`, Compose makes sure any changes made to files on your host automatically match with the corresponding files within the service container.

Sync is ideal for frameworks that support "Hot Reload" or equivalent functionality.

More generally, sync rules can be used in place of bind mounts for many development use cases.

##### Comparison to bind mounts
Compose also supports sharing a host directory inside service containers. Watch mode does not replace this functionality but exists as a companion specifically suited to developing in containers.

Most importantly, watch mode allows for greater granularity than is practical with a bind mount. Watch rules allow ignoring specific files or entire directories within the watched tree.

For example, in a JavaScript project, ignoring the `node_modules/` directory has a couple benefits:

> * **Performance**: file trees with many small files can cause high I/O load in some configurations

Reviewer: ```suggestion
* Performance: file trees with many small files can cause high I/O load in some configurations
```

> * **Multi-platform**: compiled artifacts cannot be shared if the host OS (e.g. Windows, macOS) or architecture (e.g. arm64) is different than the container

Reviewer: ```suggestion
* Multi-platform: compiled artifacts cannot be shared if the host OS (e.g. Windows, macOS) or architecture (e.g. arm64) is different than the container
```

* `target: /app/html` -> `/app/html/index.html`
* `target: /app/static` -> `/app/static/index.html`
* `target: /assets` -> `/assets/index.html`

## Example
Watch mode can be used with many different languages and frameworks.
The specific paths and rules will vary project to project, but the concepts remain the same. 

This minimal example targets a Node.js application with the following structure:
```text
myproject/
├── web/
│   ├── App.jsx
│   └── index.js
├── Dockerfile
├── compose.yaml
└── package.json
```

```yaml
services:
  web:
    build: .
    command: npm start
    x-develop:
      watch:
        - action: sync
          path: ./web
          target: /src/web
        - action: rebuild
          path: package.json
```

In this example, when running `docker compose up --build --wait`, a container for the `web` service is launched using an image built from the `Dockerfile` in the project root.
The `web` service runs `npm start` for its command, which will launch a development version of the application with Hot Module Reload enabled in the bundler (Webpack, Vite, Turbopack, etc).

After the service is up, running `docker compose alpha watch` will start watch mode.

> Then, whenever a source file in the `web/` directory is changed, Compose will sync the file to the corresponding location under `/src/web` inside the container.

Reviewer: ```suggestion
Then, whenever a source file in the `web/` directory is changed, Compose syncs the file to the corresponding location under `/src/web` inside the container.
```

> For example, `./web/App.jsx` would be copied to `/src/web/App.jsx`.

Reviewer: ```suggestion
For example, `./web/App.jsx` is copied to `/src/web/App.jsx`.
```

---
description: File watch automatically updates running services as you work
keywords: compose, file watch, experimental 
title: Automatically update services during development with Compose file watch
---

> 

Reviewer 2: The Compose v1 deprecation is missing

Watch mode monitors files and directories related to Compose services.

On change, multiple actions are available:

* **Sync**: create, overwrite, and delete corresponding files within the service container
* **Rebuild**: build a new image with BuildKit and replace the running service container

### Example `watch` configuration

```yaml
services:
  app:
    build: ./app
    x-develop:
      watch:
        - action: sync
          path: ./app/html
          target: /app/html
        - action: rebuild
          path: ./app/requirements.txt
```

## Feature stability

> This feature is still [**experimental**](/release-lifecycle/).

Reviewer 2: Maybe we should add this in a note section at the beginning of the page, wdyt?

Use `watch` to automatically update your running Compose services as you edit and save your code. 

`watch` adheres to the following file path rules:

* All paths are relative to the build context
* Directories are watched recursively
* Glob patterns are not supported
* Rules from `.dockerignore` apply
  * Use `include` / `exclude` to override
  * Temporary/backup files for common IDEs (Vim, Emacs, JetBrains, & more) are ignored automatically
* `.git` directories are ignored automatically

## Configuration

The `watch` attribute defines a list of rules that control automatic service updates based on local file changes.

Each rule requires, a `path` pattern and `action` to take when a modification is detected. There are two possible actions for `watch` and depending on
the `action`, additional fields might be accepted or required. 

### `action`

#### Sync

If `action` is set to `sync`, Compose makes sure any changes made to files on your host automatically match with the corresponding files within the service container.

Sync is ideal for frameworks that support "Hot Reload" or equivalent functionality.

> More generally, they can be used in place of bind mounts for many development use cases.

Reviewer: Is it worth expanding on this point slightly?

Author: New section JUST for comparison to bind mounts added ✨

#### Rebuild

If `action` is set to `rebuild`, Compose automatically builds a new image with BuildKit and replaces the running service container.

The behavior is the same as running `docker compose up --build <svc>`.

Rebuild is ideal for compiled languages or as fallbacks for modifications to particular files that require a full
image rebuild (e.g. `package.json`).

>**Tip**
>
> Optimize your `Dockerfile` for speedy
incremental rebuilds with [image layer caching](/build/cache)
and [multi-stage builds](/build/building/multi-stage/).
{: .tip}

### `path` and `target`

The `target` field controls how the path is mapped into the container.

For `path: ./app/html` and a change to `./app/html/index.html`:

* `target: /app/html` -> `/app/html/index.html`
* `target: /app/static` -> `/app/static/index.html`
* `target: /assets` -> `/assets/index.html`

> ## Example

Reviewer: Wonder if it isn't better to have this match the example you have in your blog? I felt your blog added some useful additional context

Author: Latest revision changes to use the same example as the blog and has explanatory text before & after
